testsuite/tests/lib-array/test_array.ml

```diff
@@ -195,3 +195,43 @@ let a = Array.init_matrix 2 3 (fun i j -> 10 * i + j);;
 [%%expect{|
 val a : int array array = [|[|0; 1; 2|]; [|10; 11; 12|]|]
 |}]
+
+let () = (* Array.equal *)
+  let test a b ~eq =
+    assert (Array.equal Int.equal a b = eq);
+    assert (Array.equal Int.equal b a = eq)
+  in
+  test [||] [||] ~eq:true;
+  test [||] [|1|] ~eq:false;
+  test [|1;2|] [|1;2|] ~eq:true;
+  test [|1;2|] [|1;2;3|] ~eq:false;
+  test [|1;2|] [|1;2;3|] ~eq:false;
+  ()
+[%%expect{|
+|}]
+
+let () = (* Array.compare *)
+  let test a b ~cmp =
+    assert (Array.compare Int.compare a b = cmp);
+    assert (Array.compare Int.compare b a = -1 * cmp);
+  in
+  test [||] [||] ~cmp:0;
+  test [||] [|0;1|] ~cmp:~-1;
+  test [|0|] [||] ~cmp:1;
+  test [|0|] [|0;1|] ~cmp:~-1;
+  test [|0;1|] [|0;1|] ~cmp:0;
+  test [|0;1|] [|0;2|] ~cmp:~-1;
+  test [|0;1|] [|0;0|] ~cmp:1;
+  test [|0;1|] [|0;1|] ~cmp:0;
+  test [|1;0|] [|0;1|] ~cmp:1;
+  test [|0;1;2|] [|0;1;2|] ~cmp:0;
+  test [|0;2;2|] [|0;1;2|] ~cmp:1;
+  test [|0;1;2|] [|0;1|] ~cmp:1;
+  (* If the length is different it is sufficent to order arrays and
+     we do not compare elements. This tests that so that
+     a possible behaviour change in the future can be detected *)
+  assert (Array.compare (fun _ _ -> assert false) [|0;1|] [|0;1;2|] = -1);
+  assert (Array.compare (fun _ _ -> assert false) [|0;1;2|] [|0;1|] = 1);
```

Comment:
The integer constants and the size difference that you use in this test are always different by exactly one unit, which mean that we cannot distinguish implementations that normalize differences to -1/0/1 and those that use negative-zero-positive results. Is this intentional? If we explicitly decided to normalize, I think we could have tests that are sensitive to it -- because user code may grow to rely on it so we don't want to change it back without noticing.

Reply:
> Is this intentional?

No, just a terrible a test case deviser. Done in 045448071eb15c12